Add Probabilistic Guard Malloc (PGM) support to bmalloc (small allocation support)

nishajain61 opened the pull request:
#### 8abe4988433e70fd4884abc737594c6b665f2e39
<pre>
Add Probabilistic Guard Malloc (PGM) support to bmalloc (small allocation support)
<a href="https://bugs.webkit.org/show_bug.cgi?id=275748">https://bugs.webkit.org/show_bug.cgi?id=275748</a>
<a href="https://rdar.apple.com/71997177">rdar://71997177</a>

Reviewed by Keith Miller.

Add Probabilistic Guard Malloc (PGM) support to bmalloc (small allocation support).

* Source/bmalloc/libpas/src/libpas/pas_deallocate.c:
(pas_try_deallocate_pgm_large):
(pas_check_pgm_entry_exists):
(pas_try_deallocate_slow_no_cache):
* Source/bmalloc/libpas/src/libpas/pas_deallocate.h:
* Source/bmalloc/libpas/src/libpas/pas_heap.h:
* Source/bmalloc/libpas/src/libpas/pas_try_allocate_common.h:
(pas_try_allocate_common_impl_slow):
* Source/bmalloc/libpas/src/libpas/pas_try_reallocate.h:
(pas_try_reallocate):
* Source/bmalloc/libpas/src/test/IsoHeapPartialAndBaselineTests.cpp:
(std::testTwoBaselinesEvictions):

Canonical link: <a href="https://commits.webkit.org/281009@main">https://commits.webkit.org/281009@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dbfa8696ad3c423887f9d2184b5c7df1e0fdb51c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/56943 "17 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/36271 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/9417 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/60563 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/7386 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/43894 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/7576 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/46121 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5186 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/58973 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/34062 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49143 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/26979 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/30842 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/6474 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/6391 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/50035 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/52802 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/6745 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/62244 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/56184 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/856 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/6850 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/53381 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/859 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49198 "Build is in progress. Recent messages:OS: Ventura (13.6.7), Xcode: 14.3; Skipping applying patch since patch_id isn't provided; Checked out pull request; Skipped layout-tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/53797 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12897 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/731 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/77945 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/32100 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/12918 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/33185 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/34270 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/32931 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

ddkilzer commented on Source/bmalloc/libpas/src/libpas/pas_deallocate.c:
```diff
@@ -64,6 +65,33 @@ bool pas_try_deallocate_known_large(void* ptr,
+        return true;
+    }
+    pas_heap_lock_unlock();
+    return false;
```
If `config_ptr->pgm_enabled` is only written once (on process launch, presumably), does it make sense to add an early return check since we wouldn't have to take the `pas_heap_lock` to check it?

(My OCD wants to combine the function calls to `pas_heap_lock_unlock()`, so I changed that as well.  Not sure if it's more CPU efficient, but should at least be more space-efficient.)

```suggestion
    if (!config_ptr->pgm_enabled)
        return false;
    pas_heap_lock_lock();
    bool result = pas_probabilistic_guard_malloc_check_exists(begin);
    if (result)
        pas_probabilistic_guard_malloc_deallocate((void *) begin);
    pas_heap_lock_unlock();
    return result;
```

nishajain61 replied:
Fixed. Thanks

ddkilzer commented on Source/bmalloc/libpas/src/libpas/pas_deallocate.c:
```diff
@@ -64,6 +65,33 @@ bool pas_try_deallocate_known_large(void* ptr,
+        return true;
+    }
+    pas_heap_lock_unlock();
+    return false;
```
This could be written a bit simpler (without a branch):

```suggestion
    pas_heap_lock_lock();
    bool result = pas_probabilistic_guard_malloc_check_exists(begin);
    pas_heap_lock_unlock();
    return result;
```

nishajain61 replied:
Fixed

ddkilzer commented on Source/bmalloc/libpas/src/libpas/pas_deallocate.c:
```diff
@@ -120,7 +148,12 @@ bool pas_try_deallocate_slow_no_cache(void* ptr,
+    /* Try to deallocate a PGM allocation based on config and checking PGM entry */
+    bool pgm_return = pas_try_deallocate_pgm_large(ptr, config_ptr);
+    if (pgm_return)
+        return true;
```
The more idiomatic way to write this in WebKit is:

```suggestion
    if (pas_try_deallocate_pgm_large(ptr, config_ptr))
        return true;
```

nishajain61 replied:
Fixed

ddkilzer commented on Source/bmalloc/libpas/src/libpas/pas_try_reallocate.h:
```diff
@@ -364,6 +363,12 @@ pas_try_reallocate(void* old_ptr,
         if (result.begin || free_mode == pas_reallocate_free_always) {
+            // Deallocate old PGM entry if its the one reallocated to other entry
+            bool pgm_return = pas_try_deallocate_pgm_large(old_ptr, config.config_ptr);
+            if (pgm_return) {
```
Same change as previous comment (don't need `pgm_return` variable):

```suggestion
            if (pas_try_deallocate_pgm_large(old_ptr, config.config_ptr)) {
```

nishajain61 replied:
Fixed

ddkilzer reviewed (commented):
I'm not really qualified to review this, but I had some minor suggestions for improvements (just small refactorings).

ddegazio reviewed (approved):
LGTM!

kmiller68 commented on Source/bmalloc/libpas/src/libpas/pas_try_allocate_common.h:
```diff
@@ -150,7 +150,33 @@ pas_try_allocate_common_impl_slow(
         break;
     }
-    
+
```
Nit: Extra new line.

kmiller68 reviewed (approved):
r=me with nit.

nishajain61 committed:
Add Probabilistic Guard Malloc (PGM) support to bmalloc (small allocation support)
https://bugs.webkit.org/show_bug.cgi?id=275748
rdar://71997177

Reviewed by Keith Miller.

Add Probabilistic Guard Malloc (PGM) support to bmalloc (small allocation support).

* Source/bmalloc/libpas/src/libpas/pas_deallocate.c:
(pas_try_deallocate_pgm_large):
(pas_check_pgm_entry_exists):
(pas_try_deallocate_slow_no_cache):
* Source/bmalloc/libpas/src/libpas/pas_deallocate.h:
* Source/bmalloc/libpas/src/libpas/pas_heap.h:
* Source/bmalloc/libpas/src/libpas/pas_try_allocate_common.h:
(pas_try_allocate_common_impl_slow):
* Source/bmalloc/libpas/src/libpas/pas_try_reallocate.h:
(pas_try_reallocate):
* Source/bmalloc/libpas/src/test/IsoHeapPartialAndBaselineTests.cpp:
(std::testTwoBaselinesEvictions):

Canonical link: https://commits.webkit.org/281009@main
